packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+  /// tree will be out of date.
+  void markNeedsSemanticsUpdate({ bool onlyChanges: false, bool noGeometry: false }) {
+    assert(!_debugDoingSemantics);
+    if (!attached || (_needsSemanticsUpdate && onlyChanges &&
+                      (noGeometry || _needsSemanticsUpdateMarkedDescendants))) {
+      return;
+    }
+    if (!noGeometry && (_semantics == null || (_semantics.hasChildren && _semantics.wasAffectedByClip))) {
+      // Since the geometry might have changed, we need to make sure to reapply any clips.
+      visitChildrenForSemantics((RenderObject child) {
+        if (!(child._needsSemanticsUpdate && child._needsSemanticsUpdateMarkedDescendants))
+          child.markNeedsSemanticsUpdate(onlyChanges: true);
+      });
+      _needsSemanticsUpdateMarkedDescendants = true;
+    }
+    if (!onlyChanges) {
+      // We might be removing the node we've been using, or adding a
+      // new one and moving our parents' children down to us. Walk the
+      // tree up to our nearest ancestor with a SemanticsNode, and
+      // mark it dirty. Since removing it would affect its parent as
+      // well, mark that dirty too. And since removing it from _that_
+      // parent might cause that parent to only have one child, we
+      // need to tell _its_ parent as well. And if that parent was
+      // already dirty, then maybe it's got some other child that's
+      // going to be removed, in which case we should tell _it's_
+      // parent too. (This step recursively.)
+      RenderObject node = this;
+      node._needsSemanticsUpdate = true;
+      node._semantics?.reset();
+      int foundSemanticsNodes = 0;
+      do {
+        if (node._semantics != null)
+          foundSemanticsNodes += 1;
+        if (node.parent is! RenderObject)
+          break;
+        node = node.parent;
+        node._needsSemanticsUpdate = true;
+        node._semantics?.reset();
+      } while (node._needsSemanticsUpdate || node._semantics == null || foundSemanticsNodes < 3);
+      _nodesNeedingSemantics.add(node);
+    } else {
+      // Either we are dirty (if we have _semantics), or our nearest
+      // ancestor with _semantics is now dirty.
+      RenderObject node = this;
+      while (node._semantics == null && node.parent is RenderObject) {
+        if (node._needsSemanticsUpdate)
+          return;
+        node._needsSemanticsUpdate = true;
+        node = node.parent;
+      }
+      if (!node._needsSemanticsUpdate) {
+        node._needsSemanticsUpdate = true;
+        _nodesNeedingSemantics.add(node);
+      }
+    }
+  }
+
+  void _updateSemantics() {
+    assert(_needsSemanticsUpdate);
+    assert(_semantics != null || parent is! RenderObject);
+    _SemanticsFragment fragment = _getSemantics();
+    assert(fragment is _InterestingSemanticsFragment);
+    SemanticsNode node = fragment.compile(parentSemantics: _semantics?.parent).single;
+    assert(node != null);
+    assert(node == _semantics);
```

Comment:
Yeah. Unfortunately I need it updated through the tree, not just at the top, so it wasn't clear how to do this without magic. (That's why I added the assert.)